File: scripts/py/qualimap_summary2csv.py
import pandas as pd
import argparse


def parse_qualimap_report(qualimapReport_txt):
    """Parse the Qualimap report text file into a dictionary."""
    with open(qualimapReport_txt, "r") as file:
        lines = file.readlines()

    out_dict = {}
    current_section = None

    for line in lines:
        line = line.strip()
        if not line:
            continue  # Skip empty lines
        if line.startswith(">>>>>>>"):
            current_section = line.strip(">>>>>>> ").strip()
            continue  # Skip section headers
        if "=" in line:
            key, value = line.split("=", 1)
            key = (
                f"{current_section} - {key.strip()}" if current_section else key.strip()
            )
            value = value.strip()
            if "(" in value:
                value, _ = value.split("(")
            if "%" in value:
                value = float(value.rstrip("%").strip().replace(",", ""))
            elif "," in value and value.replace(",", "").replace(".", "").isdigit():
                value = float(value.strip().replace(",", ""))
            else:
                value = value.strip()
            out_dict[key] = value

    return out_dict


def write_csv(data_dict, output_csv):
    """Write the parsed data dictionary to a CSV file. Replace spaces with underscores in column names."""
    out_df = pd.DataFrame([data_dict])
    out_df.columns = [col.replace(" - ", "_") for col in out_df.columns]
    out_df.columns = [col.replace("'", "") for col in out_df.columns]
    out_df.columns = [col.replace(" ", "_") for col in out_df.columns]
    # Remove commas from numeric values
    for col in out_df.columns:
        if out_df[col].dtype == object:
            out_df[col] = out_df[col].str.replace(",", "", regex=False)
    out_df.to_csv(output_csv, index=False)


def validate_file_types(input_file, output_file):
    """Ensure the input file is a .txt file and the output file is a .csv file."""
    if not input_file.endswith(".txt"):
        raise ValueError(f"Input file must be a .txt file. Provided: {input_file}")
    if not output_file.endswith(".csv"):
        raise ValueError(f"Output file must be a .csv file. Provided: {output_file}")


def parse_args():
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(
        description="Convert Qualimap report text file to CSV format."
    )
    parser.add_argument(
        "--input",
        type=str,
        required=True,
        help="Path to the Qualimap report text file.",
    )
    parser.add_argument(
        "--output", type=str, required=True, help="Path to the output CSV file."
    )
    return parser.parse_args()


def main():
    """Main function to parse Qualimap report and write to CSV."""
    args = parse_args()
    validate_file_types(args.input, args.output)
    data_dict = parse_qualimap_report(args.input)
    write_csv(data_dict, args.output)


if __name__ == "__main__":
    main()

File: scripts/R/qualimap_summary2csv.R
suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(stringr)
  library(optparse)
})

# Function to parse the Qualimap report
parse_qualimap_report <- function(qualimap_report_txt) {
  lines <- read_lines(qualimap_report_txt)
  data <- list()
  current_section <- NULL
  
  for (line in lines) {
    line <- str_trim(line)
    if (line == "") next  # Skip empty lines
    if (str_starts(line, ">>>>>>>")) {
      current_section <- str_remove(line, ">>>>>>> ") %>% str_trim()
      next
    }
    if (str_detect(line, "=")) {
      parts <- str_split(line, "=", n = 2)[[1]]
      key <- str_trim(parts[1])
      value <- str_trim(parts[2])
      if (!is.null(current_section)) {
        key <- paste(current_section, key, sep = "_")
      }
      value <- str_remove_all(value, ",")
      if (str_detect(value, "\\(")) {
        value <- str_split(value, "\\(", n = 2)[[1]][1] %>% str_trim()
      }
      if (str_detect(value, "%")) {
        value <- as.numeric(str_remove(value, "%"))
      } else if (str_detect(value, "^[0-9.]+$")) {
        value <- as.numeric(value)
      }
      data[[key]] <- value
    }
  }
  
  # Convert to data frame and clean column names
  df <- as.data.frame(data, stringsAsFactors = FALSE)
  colnames(df) <- colnames(df) %>%
    str_replace_all("\\.\\.", "_") %>%    # Replace double periods with underscores
    str_replace_all("\\.", "_") %>%    # Replace periods with underscores
    str_replace_all("[^A-Za-z0-9_]", "")  %>% # Remove special characters
    str_replace_all(" ", "_")       # Replace spaces with underscores
  
  return(df)
}

# Function to parse command-line arguments
parse_arguments <- function() {
  option_list <- list(
    make_option(
      c("--input"), 
      type = "character", 
      help = "Path to the Qualimap report text file.", 
      metavar = "FILE"
    ),
    make_option(
      c("--output"), 
      type = "character", 
      help = "Path to the output CSV file.", 
      metavar = "FILE"
    ),
    make_option(
      c("--verbose"), 
      type = "logical", 
      default = TRUE, 
      help = "Print comments during execution. Default is TRUE."
    )
  )
  
  opt_parser <- OptionParser(option_list = option_list)
  opts <- parse_args(opt_parser)
  
  return(opts)
}

# Main function
main <- function() {
  args <- parse_arguments()
  
  if (args$verbose) message("Parsing Qualimap report...")
  data <- parse_qualimap_report(args$input)
  if (args$verbose) message(glue("   Parsed data: {ncol(data)} columns"))
  
  if (args$verbose) message("Writing parsed data to CSV...")
  write_csv(data, args$output)
  
  if (args$verbose) message("Execution completed.")
}

# Run the script
if (!interactive()) {
  main()
}
